File: rl_bringup/src/toggle_setup_init.cpp
 /*
 * @name	 	  toggle_setup_init.cpp
 * @brief	 	  Simulation will be triggered for n_sec for initialize setup.
 * @details   Then, a customized move_base (https://github.com/rayvburn/navigation_flatland) is expected to orchestrate
 *            simulation steps.
 * @author  	Ronja Gueldenring
 * @date 		  2019/04/05
 **/
#include <ros/ros.h>
#include <flatland_msgs/Step.h>

// relative names will be modified to be placed within the node namespace
std::string adjustTopicName(const std::string& node_namespace, const std::string& name);

int main(int argc, char** argv){
    
  ros::init(argc, argv, "map_echo");
  ros::NodeHandle node;

  std::string train_mode_param = ros::this_node::getNamespace() + "/rl_agent/train_mode";
  int rl_mode;
  train_mode_param = adjustTopicName(ros::this_node::getNamespace(), train_mode_param);
  node.getParam(train_mode_param, rl_mode);

  bool keep_clock_running = false;
  if(rl_mode == 2){
    keep_clock_running = true;
  }
  
  float n_sec = 10.0;
  auto step_sim_srv_name = ros::this_node::getNamespace() + "/step";
  step_sim_srv_name = adjustTopicName(ros::this_node::getNamespace(), step_sim_srv_name);
  ros::ServiceClient step_simulation_ = node.serviceClient<flatland_msgs::Step>(step_sim_srv_name);

  flatland_msgs::Step msg;
  msg.request.step_time.data = 0.1;
  ros::WallTime begin = ros::WallTime::now();
  ros::WallRate r(30);
  while ((ros::WallTime::now() - begin).toSec() < n_sec || keep_clock_running) {
      if(!step_simulation_.call(msg)){
        ROS_ERROR("Failed to call step_simulation_ service");
      }
    r.sleep();
  }

  return 0;
};

std::string adjustTopicName(const std::string& node_namespace, const std::string& name) {
    if (name.length() && name.front() != '/') {
        auto name_full = "/" + node_namespace + "/" + name;
        // keep only 1 slash at the start
        std::string name_clean = "/";
        bool letter_encountered = false;
        for (const char& c : name_full) {
            if (c == '/') {
                // trim front slashes
                if (!letter_encountered) {
                    continue;
                }
                name_clean += c;
            } else {
                name_clean += c;
                letter_encountered = true;
            }
        }
        return name_clean;
    }
    return name;
};

File: rl_agent/scripts/toggle_setup_init.py
#!/usr/bin/env python
# 
# @name	 	  toggle_setup_init.cpp
# @brief	 	  Simulation will be triggered for n_sec for initialize setup.
# @details   Then, a customized move_base (https://github.com/rayvburn/navigation_flatland) is expected to orchestrate
#            simulation steps.
# @author  	Ronja Gueldenring
# @date 		  2019/04/05
#
# @rayvburn - added a Python version of `toggle_setup_init` node from `rl_bringup` (originally written in C++)
# Aims to avoid simulator dependencies during pkg compilation (to flawlessly compile on a real robot)

import rospy
import time

from flatland_msgs.srv import Step, StepRequest
from rl_agent.common_utils import adjust_topic_name

if __name__ == '__main__':
  rospy.init_node("toggle_setup_init", anonymous=True)
  train_mode_param = rospy.get_namespace() + "/rl_agent/train_mode"
  train_mode_param = adjust_topic_name(rospy.get_namespace(), train_mode_param)

  rl_mode = rospy.get_param(train_mode_param, 0)

  keep_clock_running = False
  if rl_mode == 2:
    keep_clock_running = True

  n_sec = 10.0
  step_sim_srv_name = rospy.get_namespace() + "/step"
  step_sim_srv_name = adjust_topic_name(rospy.get_namespace(), step_sim_srv_name)
  step_simulation_ = rospy.ServiceProxy(step_sim_srv_name, Step)
  step_simulation_.wait_for_service(30.0)

  msg = StepRequest()
  msg.step_time.data = 0.1
  begin = rospy.rostime.get_rostime()
  rate = rospy.Rate(10)

  # this small sleep is necessary to start rolling (otherwise rate sleeps forever)
  time.sleep(0.1)

  while (
    not rospy.is_shutdown()
    and ((rospy.rostime.get_rostime() - begin).to_sec() < n_sec or keep_clock_running)):
      try:
        if not step_simulation_.call(msg):
          rospy.logerr("Failed to call step_simulation_ service from the `toggle_setup_init` node")

        # rate.sleep()
        # There are some strange issues with rate.sleep(); when the service is called properly (returns true),
        # but the execution of the iteration consists only of the service call, the sleep blocks forever.
        # On the other hand, when there is some rospy.loginfo before the rate.sleep(), all works properly.
        # The call below acts as a workaround for the call above.
        time.sleep(0.03)
      except rospy.ROSInterruptException:
        print("Interrupt exception occurred in `toggle_setup_init.py` node")
